sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -3943,6 +3943,16 @@ object EliminateUnions extends Rule[LogicalPlan] {
   }
 }
 
+/**
+ * Removes [[Collate]] expressions if the input is already the correct type.
+ */
+object EliminateCollates extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan.transformAllExpressions {
+    case Collate(child, collation) if child.dataType.sameType(StringType(collation)) =>
```

Comment:
This seems like a nice solution, but unfortunately, in this setting we can't do it. If we just blindly remove `Collate` expressions from any place we will change a meaning of `StringType` priority. In other words, it will not have explicit meaning, but only implicit/default which is not correct, as user specifically said COLLATE in their query.

Reply:
Could we consider something like modifying flags in order to preserve priority? i.e. remove the collate expression like in this rule, but also flag the corresponding StringType as "explicit" in priority - that way we could maintain the desired priority, and also get rid of the extra COLLATE expression